registry/remote/repository.go

```diff
@@ -583,17 +595,72 @@ func (s *blobStore) Resolve(ctx context.Context, reference string) (ocispec.Desc
 	default:
 		return ocispec.Descriptor{}, errutil.ParseErrorResponse(resp)
 	}
+
+	return s.generateDescriptor(resp, refDigest)
+}
+
+// FetchReference fetches the blob identified by the reference.
+// The reference must be a digest.
+func (s *blobStore) FetchReference(ctx context.Context, reference string) (desc ocispec.Descriptor, rc io.ReadCloser, err error) {
+	ref, err := s.repo.parseReference(reference)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	refDigest, err := ref.Digest()
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	ctx = withScopeHint(ctx, ref, auth.ActionPull)
+	url := buildRepositoryBlobURL(s.repo.PlainHTTP, ref)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	resp, err := s.repo.client().Do(req)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	defer func() {
+		if err != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// no-op
+	case http.StatusNotFound:
+		return ocispec.Descriptor{}, nil, fmt.Errorf("%s: %w", ref, errdef.ErrNotFound)
+	default:
+		return ocispec.Descriptor{}, nil, errutil.ParseErrorResponse(resp)
+	}
+
+	desc, err = s.generateDescriptor(resp, refDigest)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	return desc, resp.Body, nil
```

Comment:
After checking RFC 7233, it is possible to have `httputil.ReadSeekCloser` for this method, where the request should have a range header `Range: bytes=0-`.

Reply:
Nice